Cache loaded N5 blocks for label lookup

igorpisarev opened the pull request:
This PR allows to speed up label-to-block lookups. The gain comes from caching the loaded data for other labels that are stored in the same data block.

For example, if you request `read(level, id1)` and then make a subsequent request `read(level, id2)`, it wouldn't need to re-read the data block file the second time (if the block lists for the labels `id1` and `id2` are stored in the same data block file).

This significantly improves initial 3D scene rendering performance in https://github.com/saalfeldlab/paintera/pull/267 when selecting a very large set of objects, for example, with `Ctrl+A` or `Ctrl+Shift+A`.

Commit:
cache loaded n5 blocks for label lookup

Block lists for label ids are stored in a single data block file. When requesting a list of containing blocks for a specific label id, it has to read the entire data block. Instead of ignoring the rest of the loaded data and having to read the data block file again on a subsequent query for a different label id, it now can instead cache the loaded block data. It allows to invalidate all loaded data if needed.

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -29,9 +29,9 @@ class LabelBlockLookupFromN5(
 		private const val SINGLE_ENTRY_BYTE_SIZE = 3 * 2 * java.lang.Long.BYTES
 
-		private fun fromBytes(array: ByteArray): MutableMap<Long, Array<Interval>> {
+		private fun fromBytes(array: ByteArray): TLongObjectHashMap<Array<Interval>> {
```
Why use trove map here? I don't think that the boxed `Long` values would have a negative impact here, do they?

igorpisarev replied:
I haven't done any benchmarks but thought that this wouldn't do any harm. Do you prefer `Map<Long, ..>` here because this way the key could be easily made a generic parameter?

hanslovsky replied:
I generally do not like the interface of the trove collections very much and use them only when a standard collection would inhibit performance. Generally, I think it is a good strategy to use standard collections as much as possible because most code will expect standard collections.

igorpisarev replied:
Switched back to standard `MutableMap` / `Map` collections

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -55,103 +55,125 @@ class LabelBlockLookupFromN5(
 
 	private val attributes = mutableMapOf<Int, DatasetAttributes>()
 
+	private val cache = TIntObjectHashMap<TLongObjectHashMap<TLongObjectHashMap<Array<Interval>>>>() // scale -> block index -> id -> intervals
```
Why not use an ImgLib2 cache with a custom key like
```kotlin
data class Key(
    val scale: Int,
    val blockIndex: Long,
    val id: Long)
```
?
That could also be used for invalidation, then.

igorpisarev replied:
Switched to cache key `(scale, blockId)` and some extra conversion between cache keys and input lookup keys `(scale, labelId)`

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -55,103 +55,125 @@ class LabelBlockLookupFromN5(
+		val blockId = getBlockId(level, id)
+		val map = readAndCacheMap(level, blockId)
+		map.put(id, arrayOf(*intervals))
+		writeAndCacheMap(level, blockId, map)
```
If I understand correctly: If we use imglib2 cache, we could just write and then [`Invalidate.invalidateIf`](https://github.com/imglib/imglib2-cache/blob/fe7ee81cda06a0c6bf48ae3edbd2f3c2f1067460/src/main/java/net/imglib2/cache/Invalidate.java#L70) the matching entries in the cache.

Commit:
use imglib2 cache for re-using already loaded n5 blocks with lookups

currently requires https://github.com/saalfeldlab/label-utilities/pull/6

Commit:
add autogenerated intellij files to gitignore

Commit:
fix travis build errors by setting dist: trusty

Commit:
invalidate parts of cache when setting lookups for entire scale level

igorpisarev commented:
I updated the PR with new imglib2-cache functionality. This also required some changes to the base `label-utilities` repo and currently needs its snapshot version in this PR: https://github.com/saalfeldlab/label-utilities/pull/6

Commit:
add comment about the purpose of the caching layer

Commit:
bump to label-utilities-0.5.0

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -75,83 +81,95 @@ class LabelBlockLookupFromN5(
+		cache.invalidate(blockKey)
+		val map = readBlock(blockKey).toMutableMap()
+		map[key.id] = arrayOf(*intervals)
+		writeBlock(getBlockKey(key), map)
```
Should we invalidate again after writing? To make sure that no outdated block was loaded while updating and writing.

igorpisarev replied:
Good point about concurrent access. This one is a bit tricky, so for now I just made all public methods synchronized. The access is not too frequent so it shouldn't hurt, and I tested it with a selection of ~8500k objects and didn't notice any performance implications (the difference is just a few ms for the entire set according to my benchmark).

Otherwise I find it hard to reason about how it would work without synchronization, since label ids somewhat depend on each other if they are in the same block (some individual writes may get lost, for example).

igorpisarev commented:
This one should be good to go too -- @hanslovsky could you please check and merge/release if all looks good?

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -75,83 +81,95 @@ class LabelBlockLookupFromN5(
+		cache.invalidateAll(parallelismThreshold)
 	}
 
+	override fun invalidateIf(parallelismThreshold: Long, condition: Predicate<LabelBlockLookupKey>) {
```
This can be simplified/kotlinified to
```kotlin
override fun invalidateIf(parallelismThreshold: Long, condition: Predicate<LabelBlockLookupKey>) {
	cache.invalidateIf(parallelismThreshold) { key ->
		cache.getIfPresent(key)?.keys?.any { id -> condition.test(LabelBlockLookupKey(key.level, id)) } ?: false
	}
}
```

igorpisarev replied:
Wow, that's nice -- done

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -75,83 +81,95 @@ class LabelBlockLookupFromN5(
 		return n5!!
 	}
 
+	private fun getBlockKey(key: LabelBlockLookupKey): N5LabelBlockLookupKey {
```
Kotlinify to
```kotlin
private fun getBlockKey(key: LabelBlockLookupKey) = getBlockKey(key.level, key.id)
```

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -75,83 +81,95 @@ class LabelBlockLookupFromN5(
-		LOG.debug("Reading id {} for level={}", id, level);
-		val map = readMap(level, id) ?: mutableMapOf()
-		return map.getOrElse(id, { emptyArray() })
+	override fun invalidate(key: LabelBlockLookupKey) {
```
Kotlinify to
```kotlin
override fun invalidate(key: LabelBlockLookupKey) = cache.invalidate(getBlockKey(key))
```

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -75,83 +81,95 @@ class LabelBlockLookupFromN5(
+		cache.invalidate(blockKey)
+	}
 
+	override fun invalidateAll(parallelismThreshold: Long) {
```
Kotlinify to
```kotlin
override fun invalidateAll(parallelismThreshold: Long) = cache.invalidateAll(parallelismThreshold)
```

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -75,83 +81,95 @@ class LabelBlockLookupFromN5(
+	}
+
+	private fun getBlockKey(level: Int, id: Long): N5LabelBlockLookupKey {
+		val dataset = "${String.format(scaleDatasetPattern, level)}"
```
No need to put `String.format` in quotes `""`.

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -75,83 +81,95 @@ class LabelBlockLookupFromN5(
+	@Throws(IOException::class)
+	private fun writeBlock(blockKey: N5LabelBlockLookupKey, map: Map<Long, Array<Interval>>) {
+		LOG.debug("Writing block id {} at scale level={}", blockKey.blockId, blockKey.level)
+		val dataset = "${String.format(scaleDatasetPattern, blockKey.level)}"
```
No need to put `String.format` in quotes `""`.

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -75,83 +81,95 @@ class LabelBlockLookupFromN5(
-		val attributes = this.attributes.getOrPut(level, { n5().getDatasetAttributes(dataset) })
+	private fun readBlock(blockKey: N5LabelBlockLookupKey): Map<Long, Array<Interval>> {
+		LOG.debug("Reading block id {} at scale level={}", blockKey.blockId, blockKey.level)
+		println("Reading block id " + blockKey.blockId + " at scale level " + blockKey.level)
```
Remove `println` line.

hanslovsky commented on src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt:
```diff
@@ -49,10 +56,9 @@ class LabelBlockLookupFromN5(
 			}
-
-			return map
+			return map.toMap()
```
Is it necessary to call `map.toMap()` here?

igorpisarev replied:
It's not so I went ahead and removed it -- the idea was that the cache would store immutable maps, but Map can't be modified via public interface anyway, so that's fine

hanslovsky reviewed (changes requested):
Please update PR with the suggestions. Some of them are my old code but it is good to fix them now.

Commit:
address stylistic and other minor code review comments

Commit:
synchronize all public methods

ensure predictable access and invalidation order for label ids and containing n5 blocks

Commit:
one more fix per code review comments
